### PR TITLE
Extra options for neuralyzer

### PR DESCRIPTION
**Pull request by sandergo90**
- Adding the uniqueWords faker provider with dictionary for dutch and english. When setting this method for your entity field, a unique word will be generated from the dictionary. 
- Code cleanup
- Adding the possibility to add pre_queries and post_queries to the yaml configuration file. Those queries will be executed before and after the neuralyzing process.

### CODE REVIEW

File: src/Anonymizer/DB.php [hunk 1]
```diff
@@ -230,14 +243,22 @@ private function updateData($callback = null): void
             $this->setLimit($this->dbUtils->countResults($this->entity));
         }
 
+        foreach ($this->configuration->getPreQueries() as $preQuery) {
```
Comment (sandergo90) [inline src/Anonymizer/DB.php]: @edyan they are executed here.

File: src/Configuration/ConfigDefinition.php [hunk 2]
```diff
@@ -62,6 +62,12 @@ public function getConfigTreeBuilder()
                     ->info("Faker's language, make sure all your methods have a translation")
                     ->defaultValue('en_US')
                 ->end()
+                ->arrayNode('pre_queries')
```
[the remark this answers is not in the document]
Comment (sandergo90) [reply src/Configuration/ConfigDefinition.php]: How would you implement the post and pre_action ?

File: src/Anonymizer/AbstractAnonymizer.php [hunk 3]
```diff
@@ -206,52 +212,12 @@ public function getWhereConditionInConfig(): string
         return $this->configEntites[$this->entity]['delete_where'];
     }
 
-
-    /**
-     * Generate fake data for an entity and return it as an Array
-     *
-     * @return array
-     */
-    protected function generateFakeData(): array
```
[the remark this answers is not in the document]
Comment (edyan) [reply src/Anonymizer/AbstractAnonymizer.php]: yes you can do that :)

File: src/Configuration/ConfigDefinition.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (edyan) [reply src/Configuration/ConfigDefinition.php]: I would say by callbacks or symfony expression language

Callback could be a method with parameters. Example : ['\Edyan\Neuralyzer\Utils\Db::query', 'UPDATE.....']
Problem : security

Expression langage is great as we can inject in the language our own objects, like an util. 
Look at : https://symfony.com/doc/current/components/expression_language/syntax.html#component-expression-objects

Then we could have in config : 
```yaml
pre_actions:
    - 'db.query("UPDATE....")'
```

The best would be to inject extra utils via plugins. Suppose I want to call an ftp as a pre_action, I can do my own plugin, that registers an "ftp" object with a "put" method. 

Finally, if we do that, we could even remove the delete* options. 

What do you think of that?

File: src/Anonymizer/DB.php (hunk 1 above)
Comment (edyan) [reply src/Anonymizer/DB.php]: if you agree with the previous comment, that action will be replaced by `$expressionLanguage->evaluate($preAction);`

File: src/Configuration/ConfigDefinition.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (sandergo90) [reply src/Configuration/ConfigDefinition.php]: Seems like an awesome idea actually. Do you have the time to help me out with this ? I will take a look at the previous comments today and put back the abstract functions ;)

[the remark this answers is not in the document]
Comment (sandergo90) [reply src/Configuration/ConfigDefinition.php]: Sure! What's your skype user?

[the remark this answers is not in the document]
Comment (edyan) [reply src/Configuration/ConfigDefinition.php]: emmanueldyan

[the remark this answers is not in the document]
Comment (sandergo90) [reply src/Configuration/ConfigDefinition.php]: I'm using skype for business so can't invite you. You can invite me on my email sandergo90@kodeus.onmicrosoft.com

**Comment by edyan** [comment]
Hi @sandergo90  I have pushed modifications to the branch pre_post_actions. Let's talk about it. To summarize : 
- I simplified the injection of commands / services
- Expression is util available as a service in the container, so that can be automatically injected
- It contains 2 methods : `evaluate()` and `evaluateExpressions()` (a loop that calls `evaluate()`)
- The method evaluate will work with a list of services, autoloaded too, tagged as "app.service".

Remaning : use the expression service in DB.php ... Or AbstractAnonymizer.php ! I think we can simply have a `__construct(Expression $expression)` and the job would be done :)

**Comment by sandergo90** [comment]
Hi @edyan, sorry for the late answer but your work seems very nice! Thanks for the improvements. 

So the only thing remaining would be to use the Expression service. Do you have some time to make that last change ?

**Comment by edyan** [comment]
Hi @sandergo90 I can do it next week but if you are interested to finish that, you are really welcomed !

What do you prefer ?

**Comment by edyan** [comment]
Hi @sandergo90  I have done the changes and it should work, could you please test it (see branch pre_post_actions) ?

I think it should be a great idea to remove the `delete_from` and `delete` from the config and use `pre_actions`. 


Also I have a question, why in `DB.php` you put that like that? I am afraid it could be very heavy. : 

```php
                if ($colProps['method'] === 'uniqueWord') {
                    // Count number of unique words to be taken from the dictionary.
                    $count = $this->dbUtils->countResults($this->entity);
                    $faker->addProvider(new \Edyan\Neuralyzer\Faker\Provider\UniqueWord($faker, $count, $language));
                    $faker = $faker->unique(true);
                }
```

Thanks for your review

**Comment by edyan** [comment]
Hi @sandergo90 I have refactored the uniqueWord method and added a full support for pre and post actions. I'll release that in a few days, if you have time to test, you are welcome.

**Comment by sandergo90** [comment]
Hi @edyan I did some tests and it's working perfectly! Thanks for the work! Cheers

**Comment by edyan** [comment]
Thanks @sandergo90  , I have made a new release

**Pull request closed without merging**